File: Aman/Trie/Design Add and Search Words Data Structure.cpp
class WordDictionary {
public:
    
    struct TrieNode
    {
        struct TrieNode* child[26];
        bool isEnd;
    };
    TrieNode* node;
    
    
    /** Initialize your data structure here. */
    WordDictionary() {
      node=new TrieNode();    
    }
    
    void addWord(string word) {
        TrieNode *cur=node;
        for(char c: word)
        {
           if(cur->child[c-'a']==NULL)
             cur->child[c-'a']=new TrieNode();
           cur=cur->child[c-'a'];   
        }
        cur->isEnd=true;
    }
    
    bool help(string word, TrieNode *cur)
    {
        for(int i=0;i<word.length();i++)
        {
            if(word[i]=='.')
            {
              for(int j=0;j<26;j++)
                  if(cur->child[j])
                  {
                        if(help(word.substr(i+1),cur->child[j]))
                          return true;
                  }
              return false;
            }
           else if(cur->child[word[i]-'a'] == NULL)
               return false;
           cur=cur->child[word[i]-'a'];
        }
        return cur->isEnd;
    }
    
    
    bool search(string word) {
        return help(word,node);
    }
};

File: Harish Kumar/Trie/Design Add and Search Words Data Structure/Design Add and Search Words Data Structure.java
class Node{
    
    Node child [];
    boolean isEnd;
    
    public Node(){
        child = new Node[26];
        isEnd = false;
    }
}


class WordDictionary {
    /** Initialize your data structure here. */
     Node root;
    public WordDictionary() {
       
        root = new Node();
    }
    
    public void addWord(String word) {
        
        Node curr = root;
        
        for(int i = 0 ;i < word.length();i++)
        {
            char ch = word.charAt(i);
            
            if(curr.child[ch - 'a'] == null)
            {
                curr.child[ch-'a'] = new Node();
            }
            
            curr = curr.child[ch - 'a'];
        }
        curr.isEnd = true;
    }
    
    private Boolean helper(String word, int start, Node curr)
    {
        if (curr  == null) return false;
        if(start == word.length())
        {
            if(curr.isEnd)
            {
                return true;
            }
            else
            {
                return false;
            }
        }
        
        char c = word.charAt(start);
        if(c == '.')
        {
            for(int i = 0; i < 26; i++)
            {
                if(helper(word,start+1,curr.child[i])) return true;
            }
        }
        else
        {
            if(helper(word,start+1,curr.child[c - 'a'])) return true;
        }
        
        return false;
    }
    
    public boolean search(String word) {
        
        Node curr = root;
        
        return helper(word,0, curr);
        
        
    }
}


/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */
